File: backend-legacy/src/services/image-gen.ts
/**
 * AI Artwork service.
 *
 * The model plays the artist: it authors an SVG scene inside a draw_image tool
 * call, and this module sanitizes it, saves it, and rasterizes it to a real PNG
 * (@resvg/resvg-js — bun-compatible native prebuild) so the chat UI, downloads
 * and any future vision-model review all work on a raster image.
 *
 * If rasterization is unavailable (native module missing on this machine) or
 * resvg rejects the SVG, the file is kept as SVG and the caller falls back to
 * it gracefully — generation still succeeds.
 */
import { promises as fs } from 'fs';
import path from 'path';
import crypto from 'crypto';

// Stable location for generated images. The Electron launcher sets
// GENERATED_IMAGES_DIR next to the exe so images persist across updates;
// plain `bun run src/index.ts` keeps them under <backend>/generated_images.
const GENERATED_DIR =
  process.env.GENERATED_IMAGES_DIR || path.join(process.cwd(), 'generated_images');

export const getGeneratedImagesDir = (): string => GENERATED_DIR;

const MAX_SVG_BYTES = 100 * 1024;

export interface SanitizeResult {
  ok: boolean;
  error?: string;
}

/** Reject anything that could execute or exfiltrate when the SVG is opened. */
const DANGEROUS_PATTERNS = [
  /<\s*script/i,
  /<\s*\/\s*script/i,
  /on\w+\s*=/i,
  /javascript:/i,
  /<\s*foreignobject/i,
  /<\s*iframe/i,
  /<\s*object/i,
  /<\s*embed/i,
  /<\s*image\b/i,
  /\bhref\s*=/i,
  /xlink:href/i,
];

/**
 * Validate that the model output is a single, safe SVG document.
 * Returns { ok: true } or { ok: false, error: <reason> }.
 */
export function sanitizeSvg(input: string): SanitizeResult {
  if (typeof input !== 'string') return { ok: false, error: 'svg must be a string.' };
  const svg = input.trim();
  if (!svg) return { ok: false, error: 'svg is empty.' };
  if (svg.length > MAX_SVG_BYTES) {
    return {
      ok: false,
      error: `SVG too large (${Math.round(svg.length / 1024)} KB, max ${Math.round(MAX_SVG_BYTES / 1024)} KB). Simplify the scene — fewer elements, shorter paths.`,
    };
  }

  // Allow an optional XML prolog, then exactly one <svg> root.
  const body = svg.replace(/^\s*<\?xml[^>]*\?>\s*/, '');
  const openTags = body.match(/<\s*svg[\s>]/gi) || [];
  if (openTags.length !== 1) {
    return { ok: false, error: 'The SVG must be a single document with exactly one <svg> root element.' };
  }
  const rootOpen = openTags[0];
  const closeTag = new RegExp(`<\\s*/${rootOpen.replace(/[^a-z]/gi, '')}\\s*>`, 'i');
  if (!closeTag.test(body)) {
    return { ok: false, error: 'The SVG is missing its closing </svg> tag.' };
  }

  for (const pattern of DANGEROUS_PATTERNS) {
    if (pattern.test(body)) {
      return {
        ok: false,
        error: 'SVG contains disallowed content (scripts, event handlers, external references or <image>). Only shapes, paths, gradients and defs are allowed.',
      };
    }
  }

  return { ok: true };
}

export interface SavedArtwork {
  /** The file to embed/display — ends in .png when rasterized, else .svg. */
  filename: string;
  /** Always saved SVG source (same base name as filename when .png won). */
  svgFilename: string;
  /** True when a raster PNG was produced alongside the SVG. */
  png: boolean;
}

/**
 * Rasterize SVG → PNG via resvg. Returns null (never throws) so callers can
 * fall back to serving the SVG when the native module is unavailable or the
 * SVG can't be rendered.
 */
async function rasterizeToPng(svg: string): Promise<Buffer | null> {
  let ResvgCtor: any;
  try {
    const mod = await import('@resvg/resvg-js');
    ResvgCtor = mod.Resvg;
  } catch (e) {
    console.warn('[image] resvg native module unavailable — falling back to SVG:', e instanceof Error ? e.message : String(e));
    return null;
  }

  try {
    // Render at the declared size when sane, otherwise force a 1024-wide canvas.
    const wAttr = /<svg[^>]*\bwidth=["'](\d+(?:\.\d+)?)/i.exec(svg);
    const hAttr = /<svg[^>]*\bheight=["'](\d+(?:\.\d+)?)/i.exec(svg);
    const w = wAttr ? parseFloat(wAttr[1]) : NaN;
    const h = hAttr ? parseFloat(hAttr[1]) : NaN;
    const declaredSane = Number.isFinite(w) && Number.isFinite(h) && w > 0 && h > 0 && w <= 2048 && h <= 2048;
    const options = declaredSane ? {} : { fitTo: { mode: 'width', value: 1024 } as const };

    const resvg = new ResvgCtor(svg, options);
    const png: Buffer = resvg.render().asPng();
    if (!png || png.length < 8 || png[0] !== 0x89 || png[1] !== 0x50) {
      console.warn('[image] Rasterization produced an empty/odd PNG — falling back to SVG');
      return null;
    }
    return png;
  } catch (e) {
    console.warn('[image] Rasterization failed — falling back to SVG:', e instanceof Error ? e.message : String(e));
    return null;
  }
}

/**
 * Sanitize-checked SVG in → files out. Saves `<base>.svg` always and
 * `<base>.png` when rasterization succeeds. Throws only on filesystem errors.
 */
export async function saveArtwork(svg: string): Promise<SavedArtwork> {
  await fs.mkdir(GENERATED_DIR, { recursive: true });

  const hash = crypto.createHash('md5').update(svg + Date.now()).digest('hex').substring(0, 8);
  const base = `img_${Date.now()}_${hash}`;
  const svgFilename = `${base}.svg`;
  const svgPath = path.join(GENERATED_DIR, svgFilename);
  await fs.writeFile(svgPath, svg);

  console.log(`[image] Saved SVG artwork: ${svgFilename} (${(svg.length / 1024).toFixed(1)} KB)`);

  const png = await rasterizeToPng(svg);
  if (png) {
    const pngFilename = `${base}.png`;
    await fs.writeFile(path.join(GENERATED_DIR, pngFilename), png);
    console.log(`[image] Rasterized PNG: ${pngFilename} (${(png.length / 1024).toFixed(1)} KB)`);
    return { filename: pngFilename, svgFilename, png: true };
  }

  return { filename: svgFilename, svgFilename, png: false };
}

File: backend/app/imagegen.py
"""AI Artwork service — mirrors backend/src/services/image-gen.ts.

The model plays the artist: it authors an SVG scene inside a draw_image tool
call, this module sanitizes it, saves it, and rasterizes it to a real PNG
(resvg-py). If rasterization is unavailable or resvg rejects the SVG, the file
is kept as SVG and callers fall back to it gracefully — generation still
succeeds.
"""

from __future__ import annotations

import hashlib
import re
import time
from pathlib import Path

from .config import get_generated_images_dir
from .logger import error as log_error, info as log_info, warn as log_warn

MAX_SVG_BYTES = 100 * 1024

# Reject anything that could execute or exfiltrate when the SVG is opened.
DANGEROUS_PATTERNS = [
    re.compile(r"<\s*script", re.IGNORECASE),
    re.compile(r"<\s*/\s*script", re.IGNORECASE),
    re.compile(r"on\w+\s*=", re.IGNORECASE),
    re.compile(r"javascript:", re.IGNORECASE),
    re.compile(r"<\s*foreignobject", re.IGNORECASE),
    re.compile(r"<\s*iframe", re.IGNORECASE),
    re.compile(r"<\s*object", re.IGNORECASE),
    re.compile(r"<\s*embed", re.IGNORECASE),
    re.compile(r"<\s*image\b", re.IGNORECASE),
    re.compile(r"\bhref\s*=", re.IGNORECASE),
    re.compile(r"xlink:href", re.IGNORECASE),
]


def sanitize_svg(svg: str) -> dict[str, object]:
    """Validate the model output is a single, safe SVG document.
    Returns {ok: True} or {ok: False, error: reason}."""
    if not isinstance(svg, str):
        return {"ok": False, "error": "svg must be a string."}
    svg = svg.strip()
    if not svg:
        return {"ok": False, "error": "svg is empty."}
    if len(svg.encode("utf-8")) > MAX_SVG_BYTES:
        return {
            "ok": False,
            "error": (
                f"SVG too large ({len(svg) // 1024} KB, max {MAX_SVG_BYTES // 1024} KB). "
                "Simplify the scene — fewer elements, shorter paths."
            ),
        }

    # Allow an optional XML prolog, then exactly one <svg> root.
    body = re.sub(r"^\s*<\?xml[^>]*\?>\s*", "", svg)
    open_tags = re.findall(r"<\s*svg[\s>]", body, re.IGNORECASE)
    if len(open_tags) != 1:
        return {"ok": False, "error": "The SVG must be a single document with exactly one <svg> root element."}
    root_open = open_tags[0]
    tag_name = re.sub(r"[^a-z]", "", root_open, flags=re.IGNORECASE) or "svg"
    if not re.search(rf"<\s*/\s*{tag_name}\s*>", body, re.IGNORECASE):
        return {"ok": False, "error": "The SVG is missing its closing </svg> tag."}

    for pattern in DANGEROUS_PATTERNS:
        if pattern.search(body):
            return {
                "ok": False,
                "error": (
                    "SVG contains disallowed content (scripts, event handlers, external references "
                    "or <image>). Only shapes, paths, gradients and defs are allowed."
                ),
            }
    return {"ok": True}


def _rasterize_to_png(svg: str) -> bytes | None:
    """Rasterize SVG → PNG via resvg-py. Returns None (never raises) so callers
    can fall back to serving the SVG."""
    try:
        import resvg_py  # type: ignore[import-untyped]
    except Exception as e:  # noqa: BLE001
        log_warn("[image] resvg module unavailable — falling back to SVG:", e)
        return None

    try:
        w_attr = re.search(r"<svg[^>]*\bwidth=[\"'](\d+(?:\.\d+)?)", svg, re.IGNORECASE)
        h_attr = re.search(r"<svg[^>]*\bheight=[\"'](\d+(?:\.\d+)?)", svg, re.IGNORECASE)

        def _num(m: re.Match[str] | None) -> float:
            return float(m.group(1)) if m else float("nan")

        w, h = _num(w_attr), _num(h_attr)
        declared_sane = w == w and h == h and 0 < w <= 2048 and 0 < h <= 2048

        kwargs: dict[str, object] = {}
        if not declared_sane:
            kwargs["svg_width"] = 1024
        png_bytes = resvg_py.svg_to_bytes(svg_string=svg, **kwargs)  # type: ignore[attr-defined]
        if isinstance(png_bytes, list):
            png_bytes = bytes(png_bytes)
        if (
            not png_bytes
            or len(png_bytes) < 8
            or png_bytes[0] != 0x89
            or png_bytes[1] != 0x50
        ):
            log_warn("[image] Rasterization produced an empty/odd PNG — falling back to SVG")
            return None
        return png_bytes
    except Exception as e:  # noqa: BLE001
        log_warn("[image] Rasterization failed — falling back to SVG:", e)
        return None


async def save_artwork(svg: str) -> dict[str, object]:
    """Sanitize-checked SVG in → files out. Saves <base>.svg always and
    <base>.png when rasterization succeeds. Raises only on filesystem errors."""
    gen_dir = get_generated_images_dir()
    Path(gen_dir).mkdir(parents=True, exist_ok=True)

    digest = hashlib.md5((svg + str(int(time.time() * 1000))).encode("utf-8")).hexdigest()[:8]
    base = f"img_{int(time.time() * 1000)}_{digest}"
    svg_filename = f"{base}.svg"
    svg_path = Path(gen_dir) / svg_filename
    svg_path.write_text(svg, encoding="utf-8")

    log_info(f"[image] Saved SVG artwork: {svg_filename} ({len(svg) / 1024:.1f} KB)")

    png = _rasterize_to_png(svg)
    if png is not None:
        png_filename = f"{base}.png"
        (Path(gen_dir) / png_filename).write_bytes(png)
        log_info(f"[image] Rasterized PNG: {png_filename} ({len(png) / 1024:.1f} KB)")
        return {"filename": png_filename, "svgFilename": svg_filename, "png": True}

    return {"filename": svg_filename, "svgFilename": svg_filename, "png": False}


def save_png_artwork(png: bytes, prefix: str = "img") -> str:
    """Save raw PNG bytes (e.g. from a diffusion engine) into the generated
    images dir. Returns the filename."""
    digest = hashlib.md5((str(int(time.time() * 1000)) + str(len(png))).encode("utf-8")).hexdigest()[:8]
    filename = f"{prefix}_{int(time.time() * 1000)}_{digest}.png"
    path = Path(get_generated_images_dir()) / filename
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_bytes(png)
    except Exception as e:  # noqa: BLE001
        log_error("[image] Failed to save PNG:", e)
        raise
    return filename
